src/Banner/Banner.man
.\" t
.\" @(#)Banner.1	02/07/00
.TH Banner 1.8 "Feb 7 2000" Banner
.UC
.SH NAME
\fBBanner\fP \- the AfterStep start-up banner displaying module

.SH SYNOPSIS
\fBBanner\fP [\-v | \--version] [\-h | \--help] [\fIimage\fP]
.sp
Although \fBBanner\fP can be run from the command line, it is normally 
spawned by afterstep in the \fIautoexec\fP file, in the following manner:
.nf
.sp
	Module	"I"	Banner
	Wait		"I"	Banner
.sp
.fi
or from within an afterstep menu file by 'Banner':
.nf
.sp
% cat "~/GNUstep/Library/AfterStep/start/Modules/Banner"
	Module "Banner" Banner
	MiniPixmap "mini-as.xpm"
.fi

To stop the \fBBanner\fP module, a menu item can be used:
.nf
.sp
% cat "~/GNUstep/Library/AfterStep/start/Modules/Banner_Stop"
	KillModuleByName "Stop Banner" Banner
	MiniPixmap "mini-as.xpm"
.fi

.SH DESCRIPTION
The \fBBanner\fP module displays an AfterStep Logo in the center of the
screen for 3 seconds.

.SH OPTIONS
.IP "\-h | \--help"
Print a usage message and exit successfully.

.IP "\-v | \--version"
Print version information and exit successfully.

.SH INITIALIZATION
During initialization, \fBBanner\fP will accept a command line parameter of
the pixmap to display, i.e. \fIBanner Afterstep_logo-2.xpm\fP.

.SH INVOCATION
\fBBanner\fP can be invoked from the \fIautoexec\fP file or from binding the
action "Module Banner \fI[image]\fP to a key-stroke in a \fIfeel\fP file,
or from a menu as shown in the Synopsis section above.

AfterStep will search through the directories specified in the
\fIModulePath\fP configuration in the \fIbase.#bpp\fP (where # is your
colordepth) to attempt to locate Banner.

.SH CONFIGURATION OPTIONS
\fBBanner\fP has no configuration file.

.SH AUTHOR
Robert Depenbrock <robert@eclipse.asta.uni-essen.de>
.nf
Robert Nation
.nf
David Mihm <dmihm@swbell.net> [Man page]

.SH SEE ALSO
.BR afterstep (1)


ChangeLog

2003-09-25  sasha

	* generating ChangeLog

2003-09-24  sasha

	* Completed look.Glass ; renamed PointerFore, PointerBack to
	  CursorFore, CursorBack to match old Cursor option; Moved Cursor
	  definitions from feel into look Completed Colorscheme's support
	  for Cursor's color

2003-09-23  sasha

	* Added look options for mouse pointer coloring: PointerFore and
	  PointerBack; Repaired code to update window's decorations when
	  stickiness flag is toggled;

2003-09-22  sasha

	* Added cursors to default feel file; enable cursor switching while
	  hovering over different elements of the interface; Added code to
	  change color of the cursor according to what colorscheme is set
	  to - may need to tweak a bit , or make colors configurable
	* Fixed locale parsing from LANG variable; Fixed module startup to
	  pass on current locae specified on command line

2003-09-19  sasha

	* added debugging code for cursor loading

2003-09-16  sasha

	* fixed bug with windows teleporting back to original position, if
	  moved/resized too fast

2003-09-15  sasha

	* updated configure scripts
	* Fixed configure.in to properly enable i18n in libAfterImage as
	  requested. Internationalized text should be displayed properly
	  now, even if its not in UTF8; Update src/test to work with new
	  libs and added test code for i18n testing
	* Fixed unquoted string in i18N code reported by Alex Mezin

2003-09-14  sasha

	* Added checks for locale and proper initialization of
	  libAfterImage t3ext drawing in accordance to its value; Fixed
	  problem with EdgeScroll being mangled when just look changes;
	  Fixed recently introduced bug causing wierd corners in framned
	  windows upon the look change
	* Fixed typoe for msg type in SendString being declared as some
	  wierd type of unsigned long long  - possibly causing wierdness on
	  64 bit machines

2003-09-13  sasha

	* implemented feedback functionality via sending the e-mail to
	  sasha :)

2003-09-12  sasha

	* added configure option to send e-mail back to developer on
	  installation; Fixed canvas border width handling while resizing
	  and in Pager;

2003-09-11  sasha

	* fixed handling of WM_PROTOCOL changes where WM_COMMAND would get
	  lost and session management would stop working; Found bug with
	  huge impact - all the geometry calculations were off as we were
	  forgetting border width of the frame window; Fixed handling of
	  button presses on border of the frame

2003-09-10  sasha

	* fixed NeXTish colorscheme somewhat
	* Fixed workspace state save/restore to allow terms to exec command
	  using -e, and to have argv's with spaces in them; Fixed frame
	  corner sizing to make it same width as is tbar size;

2003-09-09  sasha

	* More fixing Pager's geometry; Added -i cmd line parameter to
	  ascompose to allow for 'include' files (such as colorschemes,
	  etc.); Updated configure scripts to properly detect headers when
	  its run from libAfterBase; Updated version to beta2; Updated TEAM
	  file

2003-09-08  sasha

	* Fixed Pager to properly account for label size while sizing
	  min-desktop windows

2003-09-07  sasha

	* somewhat debugged Pager's geometry handling - need to add some
	  code to properly recalculate Pager's geometry if its soecified as
	  ratio of the main desktop vs. exact geometry

2003-09-03  sasha

	* added new AfterStep log icon; fixed wharf config file

2003-09-02  sasha

	* added new xml tag : <color> to define symbolic color names
	* changed handling of tiled images to clip it properly when
	  Right/Left align is used

2003-08-30  sasha

	* debugged workspace state handling code, and improved Wait loop

2003-08-29  sasha

	* mostly implemented new workspace state saving and restoring

2003-08-28  sasha

	* fixed workspace status generator to not save xrm params, as noone
	  seems to support thoise
	* added caching of pixmaps for unfocused bas in window decorations
	  to speed up fast focus switching
	* 64-bit fixes: CARD32 instead of Window type
	* 64-bit fixes: properties and strlen warnings
	* 64-bit fixes: converted intermodule communications to always use
	  32bit values
	* more 64 bit fixes

2003-08-27  sasha

	* 64-bit compilation fixes
	* Fixed slowness in intermodule communications manifesting itself
	  when desktops are switched
	* fixed handling of non-default MyBackgrounds, and proper adding of
	  DeskConfigs while switching desks
	* synching
	*  fixed segfaulting in Pager on look switches or after window
	  restacking
	* Fixed handling of hints flags - it was doing nothing before if
	  flags == 0 - was screwing MOtif Hints like totatlly; Fixed shaped
	  window handling in WinList; Added check to not do too much
	  redrawing when WinLIst is receiving window list - may want to do
	  the same in Pager

2003-08-26  sasha

	*  added parameters to import funcions, to allow for sized
	  image/preview stuff; Removed compression buffer from ASImage
	  structure to save up some memory; Hopefully fixed hints handling
	  on 64 bit hw

2003-08-24  sasha

	* fixed event handling when windows is getting mapped - no more
	  lost events due to not selected mask ; Fixed initial status when
	  mouse pointer is out of the screen; Fixed ICCCM size hints
	  handling to have BaseSize and MinSize interchange each other when
	  one is ommited - fixes GIMPs initial placement

2003-08-23  sasha

	* added new 3 icons to complete the set of default icons: switched
	  AfterStep to use ~/.afterstep instead of ridiculous ~/GNUstel/L/A
	  stuff

2003-08-22  sasha

	* Fixed handling of shaped windows - shape is not erroneously
	  cleared, and client's shape is subtracted from decoration's shape
	* fixed autodetection of zipped xpms and added 6 more icons to
	  default iconset

2003-08-21  sasha

	* more improvements to asimbrowser

2003-08-20  sasha

	* fixed makefile for ascp, when fltk is not found; added better
	  Style for BitchX to database; remnoved scrollbar from BitchX in
	  Wharf

src/Banner/Banner.c
/*
 * Copyright (C) 1998 Sasha Vasko <sasha at aftercode.net>
 * Copyright (C) 1996 Robert Depenbrock (robert@eclipse.asta.uni-essen.de)
 * Copyright (C) 1994 Robert Nation
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include "../../configure.h"
#ifndef XPM
int
main (int argc, char **argv)
{
  return 1;
}
#else /* XPM */
#ifdef ISC
#include <sys/bsdtypes.h>	/* Saul */
#endif

#include <stdio.h>
#include <signal.h>
#include <fcntl.h>
#include <string.h>
#include <sys/wait.h>
#include <sys/time.h>
#if defined ___AIX || defined _AIX || defined __QNX__ || defined ___AIXV3 || defined AIXV3 || defined _SEQUENT_
#include <sys/select.h>
#endif

#include <unistd.h>
#include <ctype.h>
#include <stdlib.h>

#include <X11/StringDefs.h>
#include <X11/Intrinsic.h>
#include <X11/Shell.h>
#ifdef SHAPE
#include <X11/extensions/shape.h>
#endif
#include <X11/xpm.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>

/* #ifdef BROKEN_SUN_HEADERS #include "../../include/sun_headers.h" #endif */
/* #ifdef NEEDS_ALPHA_HEADER #include "../../include/alpha_header.h" #endif */

#define MODULE_X_INTERFACE

#include "../../include/aftersteplib.h"
#include "../../include/afterstep.xpm"
#include "../../include/afterstep.h"
#include "../../include/style.h"
#include "../../include/screen.h"
#include "../../include/module.h"

typedef struct _XpmIcon
  {
    Pixmap pixmap;
    Pixmap mask;
    XpmAttributes attributes;
  }
XpmIcon;

/**************************************************************************
 * A few function prototypes
 **************************************************************************/
void RedrawWindow (void);
void GetXPMData (char **);
void GetXPMFile (char *, char *);
void change_window_name (char *str);
int flush_expose (Window w);
/* static void parseOptions (int fd[2]); */
XpmIcon view;
Window win;

char *pixmapPath = NULL;
char *pixmapName = NULL;
char *myName = NULL;

int timeout = 3000000;		/* default time of 3 seconds */

char *MyName = NULL;
Display *dpy;			/* which display are we talking to */
int screen;
ScreenInfo Scr;
int x_fd;

XSizeHints mysizehints;
Pixel back_pix, fore_pix;
GC NormalGC, FGC;
static Atom wm_del_win;

#define MW_EVENTS   (ExposureMask | ButtonReleaseMask)

/****************************************************************************
 *
 * Gets a module configuration line from afterstep. Returns NULL if there are
 * no more lines to be had. "line" is a pointer to a char *.
 * (comes from fvwm2lib)
 ***************************************************************************/

#define M_END_CONFIG_INFO    (1<<19)
#define M_CONFIG_INFO        (1<<18)

void *
GetConfigLine (int *fd, char **tline)
{
  static int first_pass = 1;
  int count, done = 0;
  static char *line = NULL;
  unsigned long header[HEADER_SIZE];

  if (line != NULL)
    free (line);

  if (first_pass)
    {
      SendInfo (fd, "Send_ConfigInfo", 0);
      first_pass = 0;
    }

  while (!done)
    {
      count = ReadASPacket (fd[1], header, (unsigned long **) &line);
      if (count > 0)
	*tline = &line[3 * sizeof (long)];
      else
	*tline = NULL;
      if (*tline != NULL)
	{
	  while (isspace (**tline))
	    (*tline)++;
	}
/*      fprintf(stderr,"%x %x\n",header[1],M_END_CONFIG_INFO); */
      if (header[1] == M_CONFIG_INFO)
	done = 1;
      else if (header[1] == M_END_CONFIG_INFO)
	{
	  done = 1;
	  if (line != NULL)
	    free (line);
	  line = NULL;
	  *tline = NULL;
	}
    }
  return 0;
}

void
do_exit (void)
{
  XDestroyWindow (dpy, win);
  XSync (dpy, 0);
  exit (0);
}

void
usage (void)
{
  printf ("Usage:\n"
	  "%s [--version] [--help] [image]\n", MyName);
  exit (0);
}

/*
 * error_handler
 * catch X errors, display the message and continue running.
 */
int
error_handler (Display * disp, XErrorEvent * event)
{
  fprintf (stderr, "%s: internal error, error code %d, request code %d, minor code %d.\n",
	 MyName, event->error_code, event->request_code, event->minor_code);
  return 0;
}

/****************************************************************************
 *
 * Creates an icon window as needed
 *
 ****************************************************************************/
int
main (int argc, char **argv)
{
  char *display_name = NULL;
  int retval = 0;
  XEvent Event;
  fd_set in_fdset;
  int fd_width;
  struct timeval value;
  int i;
  char *global_config_file = NULL;

  /* Save our program name - for error messages */
  SetMyName (argv[0]);

  i = ProcessModuleArgs (argc, argv, &(global_config_file), NULL, NULL, usage);
  if (i < argc)
    {
      pixmapName = safemalloc (strlen (argv[i]) + 1);
      strcpy (pixmapName, argv[i++]);
    }

  x_fd = ConnectX (&Scr, display_name, PropertyChangeMask);
  XSetErrorHandler (error_handler);

  fd_width = GetFdWidth ();

  /* no connection with afterstep is necessary */

  /* Get the xpm banner */
  if (pixmapName)
    GetXPMFile (pixmapName, pixmapPath);
  else
    GetXPMData (afterstep);

  /* Create a window to hold the banner */
  mysizehints.flags =
    USSize | USPosition | PWinGravity | PResizeInc | PBaseSize | PMinSize | PMaxSize;
  /* subtract one for the right/bottom border */
  mysizehints.width = view.attributes.width;
  mysizehints.height = view.attributes.height;
  mysizehints.width_inc = 1;
  mysizehints.height_inc = 1;
  mysizehints.base_height = mysizehints.height;
  mysizehints.base_width = mysizehints.width;
  mysizehints.min_height = mysizehints.height;
  mysizehints.min_width = mysizehints.width;
  mysizehints.max_height = mysizehints.height;
  mysizehints.max_width = mysizehints.width;
  mysizehints.win_gravity = NorthWestGravity;

  mysizehints.x = (Scr.MyDisplayWidth - view.attributes.width) / 2;
  mysizehints.y = (Scr.MyDisplayHeight - view.attributes.height) / 2;

  win = XCreateSimpleWindow (dpy, Scr.Root, mysizehints.x, mysizehints.y,
			     mysizehints.width, mysizehints.height,
			     1, fore_pix, None);


  /* Set assorted info for the window */
  XSetTransientForHint (dpy, win, Scr.Root);
  wm_del_win = XInternAtom (dpy, "WM_DELETE_WINDOW", False);
  XSetWMProtocols (dpy, win, &wm_del_win, 1);

  XSetWMNormalHints (dpy, win, &mysizehints);
  change_window_name ("Banner");

  XSetWindowBackgroundPixmap (dpy, win, view.pixmap);
#ifdef SHAPE
  if (view.mask != None)
    XShapeCombineMask (dpy, win, ShapeBounding, 0, 0, view.mask, ShapeSet);
#endif
  XMapWindow (dpy, win);
  XSync (dpy, 0);
  XSelectInput (dpy, win, ButtonReleaseMask);
  /* Display the window */
  value.tv_usec = timeout % 1000000;
  value.tv_sec = timeout / 1000000;
  while (1)
    {
      FD_ZERO (&in_fdset);
      FD_SET (x_fd, &in_fdset);

      if (!XPending (dpy))
#ifdef __hpux
	retval = select (fd_width, (int *) &in_fdset, 0, 0, &value);
#else
	retval = select (fd_width, &in_fdset, 0, 0, &value);
#endif
      if (retval == 0)
	do_exit ();

      if (FD_ISSET (x_fd, &in_fdset))
	{
	  /* read a packet */
	  XNextEvent (dpy, &Event);
	  switch (Event.type)
	    {
	    case ButtonRelease:
	      do_exit ();
	    case ClientMessage:
	      if (Event.xclient.format == 32 && Event.xclient.data.l[0] == wm_del_win)
		do_exit ();
	    default:
	      break;
	    }
	}
    }
  return 0;
}

/****************************************************************************
 *
 * Looks for a color XPM icon file
 *
 ****************************************************************************/
void
GetXPMData (char **data)
{
  view.attributes.valuemask = XpmReturnPixels | XpmCloseness | XpmExtensions;
  view.attributes.closeness = 40000 /* Allow for "similar" colors */ ;
  if (XpmCreatePixmapFromData (dpy, Scr.Root, data,
			       &view.pixmap, &view.mask,
			       &view.attributes) != XpmSuccess)
    {
      fprintf (stderr, "Banner: ERROR couldn't convert data to pixmap\n");
      exit (1);
    }
}

void
GetXPMFile (char *file, char *path)
{
  char *full_file = NULL;

  view.attributes.valuemask = XpmReturnPixels | XpmCloseness | XpmExtensions;
  view.attributes.closeness = 40000 /* Allow for "similar" colors */ ;

  if (file)
    full_file = findIconFile (file, path, R_OK);

  if (full_file)
    {
      if (XpmReadFileToPixmap (dpy,
			       Scr.Root,
			       full_file,
			       &view.pixmap,
			       &view.mask,
			       &view.attributes) == XpmSuccess)
	{
	  return;
	}
      fprintf (stderr, "Banner: ERROR reading pixmap file\n");
    }
  else
    fprintf (stderr, "Banner: ERROR finding pixmap file in PixmapPath\n");
  GetXPMData (afterstep);
}

void
nocolor (char *a, char *b)
{
  fprintf (stderr, "Banner: can't %s %s\n", a, b);
}

#if 0
static void
parseOptions (int fd[2])
{
  char *tline = NULL;
  int clength;

  clength = strlen (MyName);

  while (GetConfigLine (fd, &tline), tline != NULL)
    {
      if (strlen (tline) > 1)
	{
	  if (mystrncasecmp (tline,
			     CatString3 ("*", MyName, "Pixmap"),
			     clength + 7) == 0)
	    {
	      if (pixmapName == (char *) 0)
		{
		  CopyString (&pixmapName, &tline[clength + 7]);
		  if (pixmapName[0] == 0)
		    {
		      free (pixmapName);
		      pixmapName = (char *) 0;
		    }
		}
	      continue;
	    }
	  if (mystrncasecmp (tline,
			     CatString3 ("*", MyName, "Timeout"),
			     clength + 8) == 0)
	    {
	      timeout = atoi (&tline[clength + 8]) * 1000000;
	      continue;
	    }
	  if (mystrncasecmp (tline, "PixmapPath", 10) == 0)
	    {
	      CopyString (&pixmapPath, &tline[10]);
	      if (pixmapPath[0] == 0)
		{
		  free (pixmapPath);
		  pixmapPath = (char *) 0;
		}
	      else
		replaceEnvVar (&pixmapPath);
	      continue;
	    }
	}
    }
  return;
}
#endif

/**************************************************************************
 *  Change the window name displayed in the title bar.
 **************************************************************************/
void
change_window_name (char *str)
{
  XTextProperty name;

  if (XStringListToTextProperty (&str, 1, &name) == 0)
    {
      fprintf (stderr, "Banner: cannot allocate window name");
      return;
    }
  XSetWMName (dpy, win, &name);
  XSetWMIconName (dpy, win, &name);
  XFree (name.value);
}


/***********************************************************************
 *
 *  Procedure:
 *	SIGPIPE handler - SIGPIPE means is dying
 *
 ***********************************************************************/

/*ARGSUSED */
void
DeadPipe (int nonsense)
{
  exit (0);
}
#endif /* XPM */
